Generating DNSSEC keys for a brand new CDN thru the API creates DNSSEC keys for potentially the wrong delivery services and then results in a 500 internal server error

The following API results in a 500 ISE when you try to generate DNSSEC keys for a brand-spanking new CDN: 

$r->post("/api/$version/cdns/dnsseckeys/generate

Here is the error:

[2017-09-21 15:50:00,413] [ERROR] DBIx::Class::Storage::DBI::_dbh_execute(): DBI Exception: DBD::Pg::st execute failed: ERROR:  null value in column "profile" violates not-null constraint
DETAIL:  Failing row contains (null, 3399, 2017-09-21 15:50:00.391382-06). [for Statement "INSERT INTO profile_parameter ( parameter, profile) VALUES ( ?, ? ) RETURNING profile" with ParamValues: 1='3399', 2=undef] at /code/src/github.com/apache/incubator-trafficcontrol/traffic_ops/app/lib/MojoPlugins/DnssecKeys.pm line 165

This is because on this line:

https://github.com/apache/incubator-trafficcontrol/blob/master/traffic_ops/app/lib/MojoPlugins/DnssecKeys.pm#L91

If $profile_id is undef (which it will be because I have setup no traffic router servers for this CDN yet and therefore there is no profile to fetch), it will cause the following problems:

1. DNSSEC keys are created for all delivery services that have no profile regardless of CDN because of the result of this query when profile = undef

https://github.com/apache/incubator-trafficcontrol/blob/master/traffic_ops/app/lib/MojoPlugins/DnssecKeys.pm#L95

2. The error stated above because you can't create a profileparameter record with profile = undef

https://github.com/apache/incubator-trafficcontrol/blob/master/traffic_ops/app/lib/MojoPlugins/DnssecKeys.pm#L165

TLDR; 

There is a problem when profile can't be found on this line:

https://github.com/apache/incubator-trafficcontrol/blob/master/traffic_ops/app/lib/MojoPlugins/DnssecKeys.pm#L91

1. why is profile being used to decide which delivery services to create dnssec keys for? 

https://github.com/apache/incubator-trafficcontrol/blob/master/traffic_ops/app/lib/MojoPlugins/DnssecKeys.pm#L94

Each DS has a CDN so why not figure out the DS's like this

my @ds_rs = $self->db->resultset('Deliveryservice')->search( { cdn_id => cdn.id } , { prefetch => [ { 'cdn' => undef }]});

2. Can we skip the profile parameter part if no profile was ever found?

--- a/traffic_ops/app/lib/MojoPlugins/DnssecKeys.pm
+++ b/traffic_ops/app/lib/MojoPlugins/DnssecKeys.pm
@@ -86,12 +86,9 @@ sub register {
 			#add to keys hash
 			$keys->{$key} = { zsk => [@zsk], ksk => [@ksk] };
 
-			#delivery services
-			#first get profile_id
-			my $profile_id = $self->get_profile_id_by_cdn($key);
-
-			#then get deliveryservices
-			my %search = ( profile => $profile_id );
+			#find the cdn's delivery services to generate keys for
+			my $cdn = $self->db->resultset('Cdn')->find( { name => $key } );
+			my %search = ( cdn_id => $cdn->id );
 			my @ds_rs = $self->db->resultset('Deliveryservice')->search( \%search , { prefetch => [ { 'cdn' => undef }]});
 			foreach my $ds (@ds_rs) {
 				if (   $ds->type->name !~ m/^HTTP/
@@ -132,44 +129,6 @@ sub register {
 
 				#add to keys hash
 				$keys->{$xml_id} = { zsk => [@zsk], ksk => [@ksk] };
-			}
-
-			#add a param to the database to track changes
-			#check to see if param already exists
-			my $param_id = $self->db->resultset('Parameter')->search(
-				{   name        => $key . ".dnssec.inception",
-					config_file => "CRConfig.json"
-				}
-			)->get_column('id')->single();
-
-			#if exists, update
-			if ( defined($param_id) ) {
-				my $param_update = $self->db->resultset('Parameter')
-					->find( { id => $param_id } );
-				$param_update->value($inception);
-				$param_update->update();
-			}
-
-			#else insert param
-			else {
-				my $param_insert = $self->db->resultset('Parameter')->create(
-					{   name        => $key . ".dnssec.inception",
-						config_file => "CRConfig.json",
-						value       => $inception,
-					}
-				);
-				$param_insert->insert();
-				$param_id = $param_insert->id();
-
-				#insert into profile_param
-				my $pp_insert
-					= $self->db->resultset('ProfileParameter')->create(
-					{   profile   => $profile_id,
-						parameter => $param_id,
-					}
-					);
-				$pp_insert->insert();
-
 			}
 
 			my $json_data = encode_json($keys);

--- a/traffic_ops/app/lib/MojoPlugins/DnssecKeys.pm
+++ b/traffic_ops/app/lib/MojoPlugins/DnssecKeys.pm
@@ -89,7 +89,7 @@ sub register {
 			#find the cdn's delivery services to generate keys for
 			my $cdn = $self->db->resultset('Cdn')->find( { name => $key } );
 			my %search = ( cdn_id => $cdn->id );
-			my @ds_rs = $self->db->resultset('Deliveryservice')->search( \%search , { prefetch => [ { 'cdn' => undef }]});
+			my @ds_rs = $self->db->resultset('Deliveryservice')->search( \%search , { prefetch => [ 'type' ] } );
 			foreach my $ds (@ds_rs) {
 				if (   $ds->type->name !~ m/^HTTP/
 					&& $ds->type->name !~ m/^DNS/ )
@@ -100,7 +100,7 @@ sub register {
 				my $ds_id  = $ds->id;
 
 				#create the ds domain name for dnssec keys
-				my $cdn_domain_name = $ds->cdn->domain_name;
+				my $cdn_domain_name = $cdn->domain_name;
 				my $ds_name = UI::DeliveryService::get_ds_domain_name($self, $ds_id, $xml_id, $cdn_domain_name);
 
 				$self->app->log->info("Creating keys for $xml_id.");

